### PR TITLE
[AutoParallel] Support sharding stage 1 and stage 3

### PR DESCRIPTION
**Pull request by LiYuRio**
### PR types

New features

### PR changes

APIs

### Description

支持sharding stage1和stage 3

PCard-73145

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/api.py [hunk 1]
```diff
@@ -733,6 +769,139 @@ def __getattr__(self, item):
...
+                mesh=param.process_mesh,
+                placements=placements,
+            )
+        return accumulator
+
+
+class ShardingStage3Fn:
+    """
+    ShardOptimizer shard_fn for stage 3, user can use this to implement sharding stage 3
+    Args:
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_parallel/api.py]: what about mesh in __init__()

File: python/paddle/distributed/auto_parallel/api.py [hunk 2]
```diff
@@ -733,6 +769,139 @@ def __getattr__(self, item):
...
+                accumulator,
+                mesh=param.process_mesh,
+                placements=placements,
+            )
+        return accumulator
+
+
+class ShardingStage3Fn:
+    """
+    ShardOptimizer shard_fn for stage 3, user can use this to implement sharding stage 3
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_parallel/api.py]: A builtin shard_fn for ``shard_optimizer`` interface, users can pass it to ``shard_optimizer`` to implement sharding optimization with stage 3.

**Review by zhiqiu — DISMISSED** [review]
LGTM

File: python/paddle/distributed/auto_parallel/api.py (hunk 1 above)
Comment (LiYuRio) [reply python/paddle/distributed/auto_parallel/api.py]: done

File: python/paddle/distributed/auto_parallel/api.py (hunk 2 above)
Comment (LiYuRio) [reply python/paddle/distributed/auto_parallel/api.py]: done

File: python/paddle/distributed/auto_parallel/api.py [hunk 3]
```diff
@@ -733,6 +769,134 @@ def __getattr__(self, item):
         return getattr(self._inner_opt, item)
 
 
+class ShardingStage1Fn:
```
Comment (jeff41404) [inline python/paddle/distributed/auto_parallel/api.py]: Is it a strange to include `Fn` in the name of a class

File: python/paddle/distributed/auto_parallel/api.py [hunk 4]
```diff
@@ -733,6 +769,134 @@ def __getattr__(self, item):
...
+                ]
+            return shard_tensor(
+                accumulator,
+                mesh=param.process_mesh,
+                placements=placements,
+            )
+        return accumulator
+
+
+class ShardingStage3Fn:
```
Comment (jeff41404) [inline python/paddle/distributed/auto_parallel/api.py]: same issue as above

**Review by sunzhongkai588 — DISMISSED** [review]
LGTM for docs

File: python/paddle/distributed/auto_parallel/api.py (hunk 4 above)
Comment (LiYuRio) [reply python/paddle/distributed/auto_parallel/api.py]: 已去掉Fn后缀

File: python/paddle/distributed/auto_parallel/api.py (hunk 3 above)
Comment (LiYuRio) [reply python/paddle/distributed/auto_parallel/api.py]: 已去掉Fn后缀

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by sunzhongkai588 — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]
LGTM
单测超时时间设置

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**